src/theory/quantifiers/extended_rewrite.cpp

```diff
@@ -830,6 +967,81 @@ Node ExtendedRewriter::extendedRewriteEqRes(Kind andk,
   return Node::null();
 }
 
+/** sort pairs by their second (unsigned) argument */
+bool sortPairSecond(const std::pair<Node, unsigned>& a,
+                    const std::pair<Node, unsigned>& b)
+{
+  return (a.second < b.second);
+}
+
+/** A simple subsumption trie used to compute pairwise list subsets */
+class SimpSubsumeTrie
+{
+ public:
+  /** the children of this node */
+  std::map<Node, SimpSubsumeTrie> d_children;
+  /** the term at this node */
+  Node d_data;
+  /** add term to the trie
+   *
+   * This adds term c to this trie, whose atom list is alist. This adds terms
+   * s to subsumes such that the atom list of s is a subset of the atom list
+   * of c. For example, say:
+   *   c1.alist = { A }
+   *   c2.alist = { C }
+   *   c3.alist = { B, C }
+   *   c4.alist = { A, B, D }
+   *   c5.alist = { A, B, C }
+   * If these terms are added in the order c1, c2, c3, c4, c5, then:
+   *   addTerm c1 results in subsumes = {}
+   *   addTerm c2 results in subsumes = {}
+   *   addTerm c3 results in subsumes = { c2 }
+   *   addTerm c4 results in subsumes = { c1 }
+   *   addTerm c5 results in subsumes = { c1, c2, c3 }
+   * Notice that the intended use case of this trie is to add term t before t'
+   * only when size( t.alist ) <= size( t'.alist ).
+   *
+   * The last two arguments describe the state of the path [t0...tn] we
+   * have followed in the trie during the recursive call.
+   * If doAdd = true,
+   *   then n+1 = index and alist[1]...alist[n] = t1...tn. If index=alist.size()
+   *   we add c as the current node of this trie.
+   * If doAdd = false,
+   *   then t1...tn occur in alist.
+   */
+  void addTerm(Node c,
+               std::vector<Node>& alist,
```

Comment:
Why not sort `alist`? It seems like this would make the trie smaller (as far as I understand, it could be O(1) vs O(`n`!) where `n` is the length of `alist` in the most extreme case), would simplify the code, and maybe be more efficient.